### PR TITLE
Make Moment class immutable

### PR DESCRIPTION
**Pull request by cketti**
# Description
- Make `Moment` class immutable
- Use static factory methods to create `Moment` instances for better readability

This doesn't introduce any behavior changes.

### CODE REVIEW

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/utils/FahrplanMisc.java [hunk 1]
```diff
@@ -78,33 +78,22 @@ public static void addAlarm(@NonNull Context context,
...
-
-        if (startTimeInMilliSec > 0) {
-            when = startTimeInMilliSec;
-            startTime = startTimeInMilliSec;
-            moment = Moment.now();
-        } else {
-            moment = session.getStartTimeMoment();
-            startTime = moment.toMilliseconds();
-            when = moment.toMilliseconds();
-        }
```
Comment (johnjohndoe) [inline app/src/main/java/nerd/tuxmobil/fahrplan/congress/utils/FahrplanMisc.java]: Can you explain why this differentiation has been made? Is there a hint in the Git commit history which you could point to in your commit history to document this change?

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/favorites/StarredListAdapter.kt [hunk 2]
```diff
@@ -26,15 +26,9 @@ class StarredListAdapter internal constructor(
...
 
-    private val nowMoment = Moment.now()
-
     @ColorInt
     private val pastSessionTextColor = ContextCompat.getColor(context, R.color.favorites_past_session_text)
 
-    override fun initViewSetup() {
-        nowMoment.setToNow()
-    }
-
```
Comment (johnjohndoe) [inline app/src/main/java/nerd/tuxmobil/fahrplan/congress/favorites/StarredListAdapter.kt:37]: :thinking: I remember doing this here: a73a506fc054a07951036b6b5674f32293f80624. It is not documented in the Git history but I remember there was some reason to initialize the moment as early as possible. Maybe this triggers something with you.

File: commons/src/main/java/info/metadude/android/eventfahrplan/commons/temporal/Moment.kt [hunk 3]
```diff
@@ -124,6 +122,14 @@ data class Moment private constructor(var time: Instant) {
      */
     fun isBefore(moment: Moment): Boolean = time.toEpochMilli() < moment.toMilliseconds()
 
+    override fun equals(other: Any?): Boolean {
+        return time == (other as? Moment)?.time
+    }
+
+    override fun hashCode(): Int = time.hashCode()
```
Comment (johnjohndoe) [inline commons/src/main/java/info/metadude/android/eventfahrplan/commons/temporal/Moment.kt:106]: Nice! Good addition, thanks!

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/FahrplanFragment.java [hunk 4]
```diff
@@ -610,7 +610,7 @@ public void onClick(View v) {
 
     public void buildNavigationMenu() {
         Moment currentDate = Moment.now().startOfDay();
-        MyApp.LogDebug(LOG_TAG, "Today is " + currentDate.toUTCDateTime().toLocalDate());
+        MyApp.LogDebug(LOG_TAG, "Today is " + currentDate.toUtcDateTime().toLocalDate());
```
Comment (johnjohndoe) [inline app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/FahrplanFragment.java:613]: FYI: This has been [discussed earlier](https://github.com/EventFahrplan/EventFahrplan/pull/245#discussion_r376616230) and postponed. I agree it is a good step to clean up the code base and to comply with the Java language conventions. :+1: Thank you.

File: commons/src/main/java/info/metadude/android/eventfahrplan/commons/temporal/Moment.kt [hunk 5]
```diff
@@ -149,4 +123,32 @@ data class Moment(var time: Instant) {
...
+         *
+         * @param utcDate must be in ISO-8601 date format, i.e. "yyyy-MM-dd"
+         */
+        fun parseDate(utcDate: String) = Moment(LocalDate.parse(utcDate).atStartOfDay().toInstant(ZoneOffset.UTC))
+
+        /**
+         * Creates a time zone neutral [Moment] instance from the given [ZonedDateTime].
+         */
+        fun ZonedDateTime.toMoment() = Moment(this.withZoneSameInstant(ZoneOffset.UTC).toInstant())
+    }
```
Comment (johnjohndoe) [inline commons/src/main/java/info/metadude/android/eventfahrplan/commons/temporal/Moment.kt:136]: It appears a bit odd to have this extension function in the companion object but it looks like we can only have it here.

I would rephrase the comment to:

> Creates a time zone neutral [Moment] instance from this [ZonedDateTime].

File: commons/src/main/java/info/metadude/android/eventfahrplan/commons/temporal/Moment.kt [hunk 6]
```diff
@@ -149,4 +123,32 @@ data class Moment(var time: Instant) {
      * Returns true if this moment is before given [moment].
      */
     fun isBefore(moment: Moment): Boolean = time.toEpochMilli() < moment.toMilliseconds()
+
+    companion object {
+        /**
+         * Creates a time zone neutral [Moment] instance of current system clock.
+         */
+        @JvmStatic
+        fun now() = Moment(Instant.now())
```
Comment (johnjohndoe) [inline commons/src/main/java/info/metadude/android/eventfahrplan/commons/temporal/Moment.kt:115]: :+1: Thanks for omitting the unneeded parameter.

**Review by johnjohndoe — CHANGES_REQUESTED** [review]
Thank you for taking the time to cleaning up :broom: 

In general I like the changes you made - there are just a few comments I left for you to take a look.

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/utils/FahrplanMisc.java (hunk 1 above)
Comment (cketti) [reply app/src/main/java/nerd/tuxmobil/fahrplan/congress/utils/FahrplanMisc.java]: I don't care for this change. The same differentiation is made inside `Session.getStartTimeMilliseconds()`. So this is not a behavior change.

Having multiple ways to retrieve the start time of a session feels like asking for trouble. This should be fixed when cleaning up the `Session` class. It's probably a larger effort and should be out of scope for this change.

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/favorites/StarredListAdapter.kt (hunk 2 above)
Comment (cketti) [reply app/src/main/java/nerd/tuxmobil/fahrplan/congress/favorites/StarredListAdapter.kt:37]: Well, now there's no good reason left. The `Moment` was set to "now" in `SessionAdapter.getView()` shortly before `setItemContent()` was called. And the body of `setItemContent()` is the only place that accesses the extension property `Session.tookPlace` which is the only place that uses the `Moment`. So every time `setItemContent()` was called the moment was reset to the current time. With this change the behavior stays the same (give or take a couple of milliseconds).

Finding out whether or not your original method of doing this was really necessary is left as an exercise for code historians.

File: commons/src/main/java/info/metadude/android/eventfahrplan/commons/temporal/Moment.kt (hunk 5 above)
Comment (cketti) [reply commons/src/main/java/info/metadude/android/eventfahrplan/commons/temporal/Moment.kt]: > I would rephrase the comment to:
> 
> > Creates a time zone neutral [Moment] instance from this [ZonedDateTime].

Done in 6228a4d810e59226ec13a9943249f9e084e2599d.

File: commons/src/test/java/info/metadude/android/eventfahrplan/commons/temporal/MomentTest.kt [hunk 7]
```diff
@@ -107,8 +107,7 @@ class MomentTest {
     @Test
     fun isBefore() {
         val moment1 = Moment.now()
-        val moment2 = Moment.now()
-        moment2.plusSeconds(1)
+        val moment2 = Moment.now().plusSeconds(1)
```
[the remark this answers is not in the document]
Comment (cketti) [reply commons/src/test/java/info/metadude/android/eventfahrplan/commons/temporal/MomentTest.kt]: Done in 334504e8ec8934c38a3244fc5d4b135fc09cb17b.

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/utils/FahrplanMisc.java (hunk 1 above)
Comment (johnjohndoe) [reply app/src/main/java/nerd/tuxmobil/fahrplan/congress/utils/FahrplanMisc.java]: Thank you for the hint that this is duplicated within `Session.getStartTimeMilliseconds()` - I haven't seen this.

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/favorites/StarredListAdapter.kt (hunk 2 above)
Comment (johnjohndoe) [reply app/src/main/java/nerd/tuxmobil/fahrplan/congress/favorites/StarredListAdapter.kt:37]: Looks like it can be simplified. :heavy_check_mark: Successfully tested on Pixel 2, Android 10 with system date and time set back to 36C3 day 1.

_As a code historian I am always interested if the former me has left behind helpful commit messages. Never give up hoping._

**Review by johnjohndoe — APPROVED** [review]
